docs/0000-materialized-links.md

```diff
@@ -0,0 +1,187 @@
+
+- `always`: Relevant link section (links, backlinks) will be generated for all pages if they don't already exist, whether or not there are links from/to that document. The sections will be inserted at the end of the document, in order: Links, Backlinks. User may change the order of these sections in the document, and the new order will be honored.
+- `never`: (**default**): Link sections will not be generated. Existing link sections are not automatically removed.
+- `nonEmpty`: Relevant link section (links, backlinks) will be generated for all pages if they don't already exist, but only when there are links from/to that document. The sections will be inserted at the end of the document, in order: Links, Backlinks. User may change the order of these sections in the document, and the new order will be honored. If a list becomes empty by a removal of a link, the heading is not automatically removed. 
+- `updateOnly`: Relevant link sections are updated, but not created. This allows the user to opt into links and backlinks per-document basis, e.g. only generate backlinks for documents whose purpose is to act as [reference lists](https://foambubble.github.io/foam/reference-lists)
+
+#### `heading`
+
+Link headings are always generated as Markdown H2-elements, and the text is customizable using the `heading` setting.
+
+Using headings as markers has the downside that they can collide with user-defined headings (e.g. if user had a section called "Links", they would be overwritten by this feature). The default values "Foam:Links" and "Foam:Backlinks" seek to minimize the odds of such a collision, but can be overridden.
+
+### Output format
+
+Link lists are output in the following format
+
+```md
+# Sample document
+
+[[page-c]]  [[page-a]] [[page-b]] [[page-b]]
+
+<!-- One empty line before heading -->
+
+## Foam:Links
+<!-- One empty line after heading -->
+
+- [Page A](page-a.md) <!-- Title is read from linked document -->
+- [Page B](page-b.md) <!-- Links are deduplicated -->
+- [Page C](page-c.md) <!-- Links are in alphabetical order, title ascending -->
+
+## Foam:Backlinks
+<!-- One empty line after heading -->
+
+- [Page A](page-a.md) <!-- Link can appear both in forward and backlinks -->
+- [Page D](page-d.md) <!-- Links from documents -->
+
+```
+
+If document has no links or backlinks, the pertinent section is not generated.
+
+### Generating/updating lists
+
+Lists are updated as follows:
+
+- In VSCode, when file is saved (similar to link reference definitions)
+- In VSCode, When new link is added, the linked-to document is updated with the backlink
+  - Logic for background updating exists in Foam Janitor
+- In VSCode, when Janitor is executed
+- On `foam-cli janitor`
+  - Currently CLI Janitor has no way of reading workspace settings file. We'll need to add this ability
+
+The extension will replace the entire immediate content of the section following the H2-heading, either until:
+
+- Next scope (next H2 or H1 heading)
```

Comment:
When we go to implement this, I think we should be careful here. I don't know how I feel about it replacing everything until it hits the next scope. Rather than relying on something user-created (like the next H2 or H1), I would almost rather rely on something from Foam.

What if instead we put the list inside it's own fence using HTML comments similar to how `[github-markdown-toc](https://github.com/ekalinin/github-markdown-toc)` does it like so:

```markdown

Table of contents
=================

<!--ts-->
   * [gh-md-toc](#gh-md-toc)
   * [Table of contents](#table-of-contents)
   * [Installation](#installation)
   * [Usage](#usage)
      * [STDIN](#stdin)
      * [Local files](#local-files)
      * [Remote files](#remote-files)
      * [Multiple files](#multiple-files)
      * [Combo](#combo)
      * [Auto insert and update TOC](#auto-insert-and-update-toc)
      * [GitHub token](#github-token)
   * [Tests](#tests)
   * [Dependency](#dependency)
<!--te-->
```

Notice the `ts` comment for table start and `te` for table end.

Reply:
is it that easy to have a comment in md?! 🤯 
I am actually happy with either the section or comment approach, once we go for one we can use that approach across our generated code (e.g. also for link definitions)

Reply:
Right!? 

> I am actually happy with either the section or comment approach, once we go for one we can use that approach across our generated code (e.g. also for link definitions)

Agreed! Consistency is 🔑